controller/api/destination/server.go

```diff
@@ -378,23 +383,94 @@ func (s *server) sendEndpointProfile(stream pb.Destination_GetProfileServer, pod
 		// `Get` doesn't include the namespace in the per-endpoint
 		// metadata, so it needs to be special-cased.
 		endpoint.MetricLabels["namespace"] = pod.Namespace
+
+		portSpec = &policyPb.PortSpec{
+			Workload: fmt.Sprintf("%s:%s", pod.Namespace, pod.Name),
+			Port:     port,
+		}
 	}
 
-	// Send the default profile without subscribing for future updates. The
-	// profile response will also include an endpoint if the IP (or hostname)
-	// sent in the profile request maps to a pod.
+	// Send the default profile. If this is the profile for a pod, subscribe
+	// for updates from the policy server which can influence whether or not
+	// the destination is treated as opaque.
 	translator := newProfileTranslator(stream, log, "", port, endpoint)
 
-	// If there are opaque ports then update the profile translator
-	// with a service profile that has those values
+	// Send the initial destination profile so that the translator does not
+	// get blocked waiting for policy server updates. After this initial
+	// profile is sent, it will be subscribed for updates if the destination
+	// is for a pod.
 	if len(opaquePorts) != 0 {
+		// Send a destination profile update that includes the opaque
+		// ports.
 		sp := sp.ServiceProfile{}
 		sp.Spec.OpaquePorts = opaquePorts
 		translator.Update(&sp)
 	} else {
 		translator.Update(nil)
 	}
 
+	if portSpec != nil {
+		// Create a client which watches for changes to the inbound policy for
+		// portSpec.
+		client, err := s.policyClient.WatchPort(stream.Context(), portSpec)
```

Comment:
By using `stream.Context()`, when the stream shuts down we will encounter in an error in `client.Recv()` which leads to that Go routine gracefully shutting down.